### DISCUSSION TITLE
equals & hashCode를 사용하는 이유는 무엇일까?

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#33] [Java]
.

**Comment by alstn113** [comment]
equals와 hashCode 메서드를 사용하는 이유는 객체의 동등성 비교와 해시코드 생성을 위해서이다.

객체의 동등성에는 물리적 동일성과 논리적 동등성이 있다.
물리적 동일성은 == 연산자를 사용하여 두 객체의 레퍼런스(주소)가 같은지 비교한다.
논리적 동등성은 equals 메서드를 오버라이딩하여 개발자가 정의한 조건에 따라 두 객체가 논리적으로 동등한지 판단한다. 

equals메서드는 객체의 참조(메모리 주소)를 비교하지만, 일반적으로 객체의 내부 상태(멤버 변수 등)를 비교하여 동등성 여부를 판단한다.

hashCode는 객체를 해시 테이블과 같은 자료구조에 저장할 때 사용된다.
hashCode 메서드를 오버라이딩할 때는 equals메서드와 일관성을 유지해야 한다. 
두 객체가 equals 메서드로 동등하다면, hashCode 메서드도 같은 값을 반환해야 합니다.

**Comment by GIVEN53** [comment]
`equals()`는 객체의 동등성을 비교하기 위해 사용한다. 이 메서드를 오버라이딩하지 않으면 객체의 레퍼런스를 비교한다.
`hashCode()`는 객체의 해시 코드를 반환한다. 

동등성을 비교하려면 `equals()`와 `hashCode()`를 오버라이딩해야 하며, 두 객체의 해시코드가 동일하면 같은 객체로 판단된다.

**Comment by le2sky** [comment]
https://intro.f-lab.kr/insight/java-object-class-equals-hashcode?gad_source=1&gclid=CjwKCAjw_e2wBhAEEiwAyFFFo6CWQMOEGdoL4_LObXOHnHM39Z8WKYoeIraZzyXbpVh3XbplPdduwBoC7mMQAvD_BwE

**Reply by le2sky** [reply]
https://tecoble.techcourse.co.kr/post/2020-07-29-equals-and-hashCode/

**Comment by GIVEN53** [comment]
동등성 비교
해시 코드 생성
해시 코드 같음 -> equals 같음 -> 같은 객체
